File: src/main/java/com/marsraver/WledFx/wled/model/WledInfo.java
package com.marsraver.WledFx.wled.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Data;

import java.util.List;
import java.util.Map;

@Data
public class WledInfo {
    // IP address (added when discovered, not in JSON response)
    private String ip;
    
    // General information
    private String ver; // Version name
    private long vid; // Build ID
    private String name; // Friendly name of the light
    private int udpport; // UDP port for realtime packets and WLED broadcast
    private boolean live; // Receiving realtime data via UDP or E1.31
    private String lm; // Realtime data source
    private String lip; // Realtime data source IP address
    private int ws; // Number of connected WebSockets clients
    private int fxcount; // Number of effects included
    private int palcount; // Number of palettes configured
    private String arch; // Platform name
    private String core; // Version of the underlying SDK
    private int lwip; // Version of LwIP
    private long freeheap; // Bytes of heap memory available
    private long uptime; // Time since last boot/reset in seconds
    private int opt; // Used for debugging purposes
    private String brand; // Producer/vendor of the light
    private String product; // Product name
    private String mac; // Hardware MAC address
    
    // Additional fields from actual device
    private String cn; // Device codename
    private String release; // Release name
    private boolean str; // Streaming enabled
    private boolean simplifiedui; // Simplified UI enabled
    private int liveseg; // Live segment ID
    private int cpalcount; // Custom palette count
    private List<MapInfo> maps; // List of mappings
    private int clock; // CPU clock speed in MHz
    private int flash; // Flash size in MB
    private String time; // Current time
    private Map<String, Object> u; // UI data
    
    // LED setup information
    @JsonProperty("leds")
    private LedsInfo leds;
    
    // WiFi information
    @JsonProperty("wifi")
    private WifiInfo wifi;
    
    // Filesystem information
    @JsonProperty("fs")
    private FsInfo fs;
    
    // Node discovery count
    private int ndc; // Number of other WLED devices discovered on the network
    
    @Data
    public static class LedsInfo {
        private boolean cct; // Supports color temperature control
        private int count; // Total LED count
        private int fps; // Current frames per second
        private boolean rgbw; // LEDs are 4-channel (RGB + White)
        private boolean wv; // White channel slider should be displayed
        private int pwr; // Current LED power usage in milliamps
        private int maxpwr; // Maximum power budget in milliamps
        private int maxseg; // Maximum number of segments supported
        private int lc; // Logical AND of all active segment's virtual light capabilities
        private List<Integer> seglc; // Per-segment virtual light capabilities
        private int bootps; // Boot preset
        private MatrixInfo matrix; // Matrix configuration
    }
    
    @Data
    public static class MatrixInfo {
        private int w; // Matrix width
        private int h; // Matrix height
    }
    
    @Data
    public static class MapInfo {
        private int id; // Mapping ID
    }
    
    @Data
    public static class WifiInfo {
        private String bssid; // BSSID of the connected network
        private int signal; // Relative signal quality
        private int channel; // Current WiFi channel
        private int rssi; // Received signal strength indicator
        private boolean ap; // Access point mode
    }
    
    @Data
    public static class FsInfo {
        private long u; // Used filesystem space in kilobytes
        private long t; // Total filesystem size in kilobytes
        private long pmt; // Last modification timestamp of the presets.json file
    }
}

File: src/main/kotlin/com/marsraver/wledfx/wled/model/WledInfo.kt
package com.marsraver.wledfx.wled.model

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty

@JsonIgnoreProperties(ignoreUnknown = true)
data class WledInfo(
    var ip: String? = null,
    var ver: String? = null,
    var vid: Long = 0,
    var name: String? = null,
    var udpport: Int = 0,
    var live: Boolean = false,
    var lm: String? = null,
    var lip: String? = null,
    var ws: Int = 0,
    var fxcount: Int = 0,
    var palcount: Int = 0,
    var arch: String? = null,
    var core: String? = null,
    var lwip: Int = 0,
    var freeheap: Long = 0,
    var uptime: Long = 0,
    var opt: Int = 0,
    var brand: String? = null,
    var product: String? = null,
    var mac: String? = null,
    var cn: String? = null,
    var release: String? = null,
    var str: Boolean = false,
    var simplifiedui: Boolean = false,
    var liveseg: Int = 0,
    var cpalcount: Int = 0,
    var maps: List<MapInfo> = emptyList(),
    var clock: Int = 0,
    var flash: Int = 0,
    var time: String? = null,
    var u: Map<String, Any?> = emptyMap(),
    @JsonProperty("leds")
    var leds: LedsInfo? = null,
    @JsonProperty("wifi")
    var wifi: WifiInfo? = null,
    @JsonProperty("fs")
    var fs: FsInfo? = null,
    var ndc: Int = 0,
) {
    @JsonIgnoreProperties(ignoreUnknown = true)
    data class LedsInfo(
        var cct: Boolean = false,
        var count: Int = 0,
        var fps: Int = 0,
        var rgbw: Boolean = false,
        var wv: Boolean = false,
        var pwr: Int = 0,
        var maxpwr: Int = 0,
        var maxseg: Int = 0,
        var lc: Int = 0,
        var seglc: List<Int> = emptyList(),
        var bootps: Int = 0,
        var matrix: MatrixInfo? = null,
    )

    @JsonIgnoreProperties(ignoreUnknown = true)
    data class MatrixInfo(
        var w: Int = 0,
        var h: Int = 0,
    )

    @JsonIgnoreProperties(ignoreUnknown = true)
    data class MapInfo(
        var id: Int = 0,
    )

    @JsonIgnoreProperties(ignoreUnknown = true)
    data class WifiInfo(
        var bssid: String? = null,
        var signal: Int = 0,
        var channel: Int = 0,
        var rssi: Int = 0,
        var ap: Boolean = false,
    )

    @JsonIgnoreProperties(ignoreUnknown = true)
    data class FsInfo(
        var u: Long = 0,
        var t: Long = 0,
        var pmt: Long = 0,
    )
}

